Mock metadata aggregation server

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1597

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

-  64052c5449d21ef3002d6ddf21dc06669e4c9315
  :round_pushpin: **Basic skeleton for refreshing metadata.**
  Still requires proper log messages and error handling. But this already allows some level of testing.

- 7be9a83dd78c02cef31b1ba71d08c5838709e98c
  :round_pushpin: **log events while refreshing metadata**
  
- 566d95d9482dd32eb4098288ecca99ac994bb36f
  :round_pushpin: **handle 404 in the metadata client directly**
  So that the calling code needs not to worry about the 'ClientError'. Any error becomes unexpected and all errors can be treated consistently.

- 398bb7839e105fabeea062292d6cbfe40e01ed8b
  :round_pushpin: **merge 'refresh' and 'Client' so that the client is already caching results by itself.**
  This makes it slightly easier to use from a consumer perspective. Once one has a handle on a 'Client IO Api' it can makes requests without having to
worry about whether they get cached or not. Doing so, I also removed the 'ClientError' from the response. The client does handle error by itself and
log messages accordingly. A caller would likely not do anything more than relogging the error.
Plus, from the caller perspective, a metadata not being there and the request failing has pretty much the same end: no metadata.

- 98a499d1e5fdbf20361074a6fe30be7766747852
  :round_pushpin: **write small unit / property tests to exercise the client.**
    Written as property so we get some fuzzy testing and test cases with and without metadata
  Fails nicely with each log messages obtained in the run:
  ```
  Assertion failed (after 1 test):
    PoolId {getPoolId = "\SOH\NUL\SOH\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\NUL\SOH\SOH\SOH\SOH\NUL\SOH\NUL\SOH\SOH\SOH\NUL\SOH\NUL\SOH\NUL\NUL\NUL\SOH\SOH\NUL"}
    MsgRefreshingMetadata (PoolId {getPoolId = "\SOH\NUL\SOH\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\NUL\SOH\SOH\SOH\SOH\NUL\SOH\NUL\SOH\SOH\SOH\NUL\SOH\NUL\SOH\NUL\NUL\NUL\SOH\SOH\NUL"}) (Just (StakePoolOffChainMetadata {ticker = StakePoolTicker {unStakePoolTicker = "VUHQ"}, name = "_-NfdG\21785\&0?-wjm'\\.G{0MtJ4E", description = "_-NfdG\21785\&0?-wjm'\\.G{0MtJ4E", homepage = "https://\tXuP9h>v\ETXXM\FSUHd\ACKp\SOH\FS\SOHg\DEL$.io"}),2020-05-04 23:31:26.917183969 UTC)
    MsgUsingCached (PoolId {getPoolId = "\SOH\NUL\SOH\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\NUL\SOH\SOH\SOH\SOH\NUL\SOH\NUL\SOH\SOH\SOH\NUL\SOH\NUL\SOH\NUL\NUL\NUL\SOH\SOH\NUL"}) 2020-05-04 23:31:26.917183969 UTC
  ```

- 51cf00ad9c6391dcc1a600298150c35909b1cd3f
  :round_pushpin: **add an extra scenario illustrating a unhappy path**
    ```
  Cardano.Pool.Metadata
    Metadata - MockServer
      Mock Server works as intended
        +++ OK, passed 100 tests:
        51% Got Valid Metadata
        49% No Corresponding Metadata
      Cache metadata when called twice within the TTL
        +++ OK, passed 10 tests.
      Fetch them again when fetching outside of the TTL
        +++ OK, passed 10 tests.
      Returns 'Nothing' and a warning log message on failure
        +++ OK, passed 1 test.
  ```

- c76f29d7b3eb338e26889d9ec41f8ecb37cd08d3
  :round_pushpin: **move 'count' to 'Test.Util.Trace' with extra comments'**

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
Basic skeleton for refreshing metadata.

Still requires proper log messages and error handling. But this already allows some level of testing.

KtorZ committed:
log events while refreshing metadata

KtorZ committed:
handle 404 in the metadata client directly

So that the calling code needs not to worry about the 'ClientError'. Any error becomes unexpected and all errors can be treated consistently.

Anviking commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -65,23 +95,123 @@ type Api
+    { saveMetadata
+        :: PoolId -> (Maybe StakePoolOffChainMetadata, UTCTime) -> m ()
+        -- ^ A callback action for storing an off-chain metadata. The callback
+        -- may be called with 'Nothing' to store that no metadata were found for
```
> may be called with 'Nothing' to store that no metadata were found

Does this mean any existing metadata will be removed?

> this allows for not constantly re-fetching data for pools that are known to have no metadata

Not quite understanding this from the comment

KtorZ replied:
> Does this mean any existing metadata will be removed?

No. This would be done automatically by the DB when pushing a de-registration certificate.

> Not quite understanding this from the comment

The thing is: we don't want to only cache metadata when they exist, but also when they aren't available. If we only cache metadata hit, then everytime we'll "refresh" metadata we'll have to make a query for all the pools that didn't have metadata. 

Yet, if we cache both the presence or absence of metadata, we only need a request for pools that are new and for which the state is unknown.

Anviking commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -65,23 +95,123 @@ type Api
+    -> [PoolId]
+        -- ^ A list of pool ids for which metadata need to be fetched or
+        -- refreshed.
+    -> IO ()
```
Why is `refresh` needed? Can we not just put the caching inside `Client`?
```haskell
cachedClient 
    :: (PoolId -> m (Maybe StakePoolOffChainMetadata, UTCTime)) 
    -> (PoolId -> Maybe (StakePoolOffChainMetadata, UTCTime) -> m ()) 
    -> Client Api m 
    -> Client Api m
cachedClient pid putCached getCached rawClient = Client { getStakePoolMetadata
    if stale 
    then (getStakePoolMetadata rawClient) >>= \x -> putCache x >> return x 
    else getCached
}
```

KtorZ replied:
That is a valid point. Felt simpler to write both decoupled initially but, simply having a cached client make sense.

Anviking replied:
I was more curious about a `Client Api m -> Client Api m` function, but sure, what you have now works too.

KtorZ committed:
merge 'refresh' and 'Client' so that the client is already caching results by itself.

This makes it slightly easier to use from a consumer perspective. Once one has a handle on a 'Client IO Api' it can makes requests without having to
worry about whether they get cached or not. Doing so, I also removed the 'ClientError' from the response. The client does handle error by itself and
log messages accordingly. A caller would likely not do anything more than relogging the error.
Plus, from the caller perspective, a metadata not being there and the request failing has pretty much the same end: no metadata.

KtorZ committed:
write small unit / property tests to exercise the client.

  Written as property so we get some fuzzy testing and test cases with and without metadata
  Fails nicely with each log messages obtained in the run:
  ```
  Assertion failed (after 1 test):
    PoolId {getPoolId = "\SOH\NUL\SOH\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\NUL\SOH\SOH\SOH\SOH\NUL\SOH\NUL\SOH\SOH\SOH\NUL\SOH\NUL\SOH\NUL\NUL\NUL\SOH\SOH\NUL"}
    MsgRefreshingMetadata (PoolId {getPoolId = "\SOH\NUL\SOH\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\NUL\SOH\SOH\SOH\SOH\NUL\SOH\NUL\SOH\SOH\SOH\NUL\SOH\NUL\SOH\NUL\NUL\NUL\SOH\SOH\NUL"}) (Just (StakePoolOffChainMetadata {ticker = StakePoolTicker {unStakePoolTicker = "VUHQ"}, name = "_-NfdG\21785\&0?-wjm'\\.G{0MtJ4E", description = "_-NfdG\21785\&0?-wjm'\\.G{0MtJ4E", homepage = "https://\tXuP9h>v\ETXXM\FSUHd\ACKp\SOH\FS\SOHg\DEL$.io"}),2020-05-04 23:31:26.917183969 UTC)
    MsgUsingCached (PoolId {getPoolId = "\SOH\NUL\SOH\NUL\NUL\NUL\NUL\NUL\NUL\SOH\SOH\NUL\SOH\SOH\SOH\SOH\NUL\SOH\NUL\SOH\SOH\SOH\NUL\SOH\NUL\SOH\NUL\NUL\NUL\SOH\SOH\NUL"}) 2020-05-04 23:31:26.917183969 UTC
  ```

KtorZ committed:
add an extra scenario illustrating a unhappy path

  ```
  Cardano.Pool.Metadata
    Metadata - MockServer
      Mock Server works as intended
        +++ OK, passed 100 tests:
        51% Got Valid Metadata
        49% No Corresponding Metadata
      Cache metadata when called twice within the TTL
        +++ OK, passed 10 tests.
      Fetch them again when fetching outside of the TTL
        +++ OK, passed 10 tests.
      Returns 'Nothing' and a warning log message on failure
        +++ OK, passed 1 test.
  ```

rvl commented on lib/core/test/unit/Cardano/Pool/MetadataSpec.hs:
```diff
@@ -237,3 +321,14 @@ generateWith seed (MkGen gen) =
+--
+-- >>> count (_Ctor @"MyConstructor") xs
+-- 2
+count :: Prism' s a -> [s] -> Int
```
Maybe put this in a tests common module, because other tests which count log messages can use it.

KtorZ replied:
c76f29d7b3eb338e26889d9ec41f8ecb37cd08d3

KtorZ committed:
move 'count' to 'Test.Util.Trace' with extra comments'

Commit:
Regenerate nix

Anviking commented on lib/core/test/unit/Cardano/Pool/MetadataSpec.hs:
```diff
@@ -83,54 +93,120 @@ import Test.QuickCheck
+                void $ getStakePoolMetadata pid
+                void $ getStakePoolMetadata pid
+            monitor $ counterexample $ unlines $ show <$> logs
+            assert $ countMsg (_Ctor @"MsgRefreshingMetadata") logs == 1
```
Cool approach!

Anviking commented on lib/core/test/unit/Cardano/Pool/MetadataSpec.hs:
```diff
@@ -83,54 +93,120 @@ import Test.QuickCheck
+                threadDelay' (2 * defaultCacheTTL)
+                void $ getStakePoolMetadata pid
+            assert $ countMsg (_Ctor @"MsgRefreshingMetadata") logs == 2
+            assert $ countMsg (_Ctor @"MsgUsingCached")        logs == 0
```
But now, what if the data becomes unavailable? How should it react?

KtorZ replied:
If the metadata is removed, then it'll be updated when the cache is refreshed.

KtorZ committed:
relax cacheTTL delay in testing, makes it larger for CI & windows.
